Improve error and status handling

CaselIT opened the pull request:
# Summary of Changes

Improve error and status handling:
- all falcon errors have a representation
- remove `has_representation` and `OptionalRepresentation`
- ensure that falcon clears the previous body when handling an `HttpStatus` or an `Error`
- fix `HttpError.to_xml` return type docs and set it directly to `Request.data`

# Related Issues

Fixes #452

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [x] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [x] Added **tests** for changed code.
- [x] Prefixed code comments with GitHub nick and an appropriate prefix.
- [x] Coding style is consistent with the rest of the framework.
- [x] Updated **documentation** for changed code.
    - [x] Added docstrings for any new classes, functions, or modules.
    - [x] Updated docstrings for any modifications to existing code.
    - [x] Updated both WSGI and ASGI docs (where applicable).
    - [x] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [x] Updated all relevant supporting documentation files under `docs/`.
    - [x] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [x] Changes (and possible deprecations) have [towncrier](https://towncrier.readthedocs.io/en/actual-freaking-docs/index.html) news fragments under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

*PR template inspired by the attrs project.*

CaselIT commented:
not sure why codecov misreported the coverage

kgriffs commented:
Blocked on: https://github.com/pytest-dev/pytest-randomly/issues/246

If a new version of pytest-randomly is not immediately forthcoming, we'll have to pin the version.

CaselIT commented:
that was fast. already released https://github.com/pytest-dev/pytest-randomly/issues/246#issuecomment-614183962

vytas7 commented on tests/test_httperror.py:
```diff
@@ -583,15 +583,15 @@ def test_405_without_body(self, client):
         response = client.simulate_request(path='/405')
         assert response.status == falcon.HTTP_405
-        assert not response.content
+        assert response.text == falcon.HTTPMethodNotAllowed(['PUT']).to_json()
```
Do we have assertions against the actual response content?
The above approach is cool and DRY, but if there was a bug in `to_json()`, this would pass. In other words, this test is just verifying that the `to_json()` method is called, and the call result is transferred to the response body.

CaselIT replied:
Yes, this test does only that. It assumes that `.to_json()` is correct.
`.to_json()` should be already tested somewhere else. 

In case it's missing I can add that test, but I would not retest it here, or this test (all the other similar ones) become overly broad

vytas7 replied:
As long as `to_json()` is tested elsewhere for `HTTPMethodNotAllowed`, it's fine.

CaselIT replied:
It seems not. I'll test in one tests

vytas7 commented on docs/_newsfragments/777.breakingchange.rst:
```diff
@@ -1,2 +1,2 @@
-The class :class:`~.falcon.http_error.NoRepresentation` was removed since
-all :class:`~.falcon.HTTPError` have optional representation.
\ No newline at end of file
+The class :class:`~.falcon.http_error.NoRepresentation` was removed. All
```
Same as above, the breaking change is too abrupt IMHO.

CaselIT replied:
This was removed in https://github.com/falconry/falcon/pull/1655

CaselIT replied:
corrected to deprecated, since it was readded

vytas7 commented on falcon/app.py:
```diff
@@ -880,6 +870,9 @@ def _compose_status_response(self, req, resp, http_status):
         # it's acceptable to set resp.body to None (to indicate no body).
         resp.body = http_status.body
+        # NOTE(caselit): Reset the data and media, so that the body is
+        # sent in all cases, even when it's None
```
Could explain the "even when it's `None`" part? How can the body be sent if it is `None`?

CaselIT replied:
I'll reword it.
What I meant it that we should not use data or media of body is `None`

CaselIT commented on falcon/http_error.py:
```diff
@@ -136,10 +123,6 @@ def __init__(self, status, title=None, description=None, headers=None,
     def __repr__(self):
         return '<%s: %s>' % (self.__class__.__name__, self.status)
 
-    @property
```
Ok.

CaselIT commented on falcon/http_error.py:
```diff
@@ -208,26 +191,3 @@ def to_xml(self):
                 et.tostring(error_element, encoding='utf-8'))
-
-
-class OptionalRepresentation:
```
In https://github.com/falconry/falcon/pull/1655 `NoRepresentation` was removed completely. I think that we should keep both or remove both

vytas7 replied:
Agreed that we should be consistent.

My vote, as also explained in other comments, is on keeping both, as well as keeping `has_representation`.
@kgriffs thoughts?

CaselIT commented:
I've update the code after the review. The only thing left to decide is what to do regarding
`OptionalRepresentation ` and `NoRepresentation `

vytas7 commented on falcon/http_error.py:
```diff
@@ -138,6 +127,8 @@ def __repr__(self):
     @property
     def has_representation(self):
+        warn = 'has_representation is deprecated and is currently unused by falcon'
+        warnings.warn(warn, DeprecatedWarning)
```
In the default CPython configuration, these are unfortunately oftentimes suppressed.

Since under the hood a property is a function, maybe you could just reuse our [`deprecated`](https://github.com/falconry/falcon/blob/master/falcon/util/misc.py#L123) utility?

CaselIT replied:
I use `DeprecatedWarning` of falcon.

I'm not using deprecated since it adds `Call to deprecated function has_representation(...)` before the specified warning, and I thought that it may create confusion

kgriffs commented:
I agree that we should either remove both classes or keep both. That being said, if we leave them around and just say they are deprecated (and the framework itself no longer makes use of them; that could still be a breaking change that we leave in and note in the news fragment, since it should have minimal impact), we would have to put back the checks for `has_representation` in case someone's custom error class is using one of the deprecated mixins. 

By itself, removing the classes altogether would probably not cause a lot of issues for existing apps, but I can appreciate @vytas7 "thousand paper cuts" argument. Perhaps it would be a good idea to avoid introducing additional breaking changes going forward for the remainder of the 3.x work.

CaselIT commented:
While I agree that breaking changes should be minimized, I don't think there are drastic changes for version 3 so I don't think those will cause a lot of issues.
And in the current milestone for v3 of breaking changes there is only this one and #1589.

That said, I'm cool either way

> we would have to put back the checks for has_representation in case someone's custom error class is using one of the deprecated mixins.

I don't think we should restore it even if we re-add the mixin classes.
It's easy to provide a custom serializer with the previous logic if one is needed:

```py
def mySerializer(req, resp, exception):
    if not exception.has_representation:
        return
    return falcon.app_helpers.default_serialize_error(req,resp,expeption)
```

vytas7 commented:
Good points @kgriffs @CaselIT ... Right, this is a bit tricky, because it's hard to make the change without actually making a change :thinking:

I'm not too opinionated about this.
Just off the top of my head, I would probably do something along the lines of
* Resurrect `NoRepresentation`, keep `OptionalRepresentation`. Document them as deprecated.
* Keep the `has_representation` property, but document it as deprecated. Referencing emits a warning.
* **Breaking change** (but very low impact). No standard Falcon classes any longer use these mixins.
* **Breaking change**. The default error serializer no longer respects the `has_representation` property.

CaselIT commented:
If we all agree can update this with the proposed changes in the latest message.

Regarding NoRepresentation and OptionalRepresentation, since they are mixins, should I add a warning? The best place for this, would probably be [__init_subclass__](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) since they don't have __init__ and it's probably best to warn when a subclass is defined on application load, but that's only py3.6+.
I can fix it with a metaclass, but I'm not sure it's worth the hassle for a deprecated feature...

Suggestions?

vytas7 commented:
I don't think we need any deprecation warnings on mixins, as there is no clean way indeed.
The only meaningful logic these mixins provide are centered on the `has_representation` property anyway, so as long as referencing it emits a warning, we should be fine IMHO.

CaselIT commented:
Ok, I will warn in `has_representation`

CaselIT commented:
I've updated the pull request.

I've added an attribute to deprecated for when it's used on properties. I've also removed the exception on warning when falcon testing is the env, since I think we should test that the deprecated stuff actually raises the deprecation warning in the tests.

CaselIT commented:
MyPy fails in a strange way.
> falcon/http_error.py:126: error: Decorated property not supported
> falcon/http_error.py:227: error: Decorated property not supported
> falcon/http_error.py:259: error: Decorated property not supported
> Found 3 errors in 1 file (checked 77 source files)

is there a way of skipping these checks?

kgriffs commented:
> MyPy fails in a strange way.
> 
> > falcon/http_error.py:126: error: Decorated property not supported
> > falcon/http_error.py:227: error: Decorated property not supported
> > falcon/http_error.py:259: error: Decorated property not supported
> > Found 3 errors in 1 file (checked 77 source files)
> 
> is there a way of skipping these checks?

You can try appending `# type: ignore` to the lines in question.

CaselIT commented:
Thansk, that did it

CaselIT commented:
Anything missing from this one?

vytas7 commented on docs/_newsfragments/452.breakingchange.rst:
```diff
@@ -0,0 +1,6 @@
+serializer now generates a representation for every error type that derives from
+:class:`falcon.HTTPError`. In addition, Falcon now ensures that any previously
+set response body is cleared before handling any subclass of
+:class:`falcon.HTTPStatus` or any other raised exception.
```
Not sure the latter part of the statement is universally true.

Consider, for instance, the following app:
```python
import falcon


class Resource:

    def on_get(self, req, resp):
        resp.media = {'message': 'Hello, World!'}
        resp.set_header('X-Value', str(0/0))


def handle_zero_division(req, resp, ex, params):
    print('resp.media ==', resp.media)
    print('This went wrong:', ex)


app = falcon.App()
app.add_error_handler(ZeroDivisionError, handle_zero_division)
app.add_route('/test', Resource())
```
```
resp.media == {'message': 'Hello, World!'}
This went wrong: division by zero
```
I.e., it seems that when handling _any other_ raised exception, the response body is not necessarily cleared.

CaselIT replied:
Nice catch. I'll fix that, since my intention was to do as in the note :)

vytas7 commented on falcon/http_error.py:
```diff
@@ -210,6 +201,39 @@ def to_xml(self):
+        `has_representation` property as expected.
+
+    Warning:
+        As of Falcon 3.0, this mixin class is no longer used since all falcon
```
_Falcon_ and _falcon_ -- different capitalization in the same line.

CaselIT commented:
Updated the documentation and fixed the bug on custom exceptions.
@vytas7 can you take another look

vytas7 commented on falcon/app.py:
```diff
@@ -936,6 +926,8 @@ def _handle_exception(self, req, resp, ex, params):
         err_handler = self._find_error_handler(ex)
 
+        # NOTE(caselit): Reset body, data and media before calling the handler
+        resp.body = resp.data = resp.media = None
```
Shouldn't we move these statements inside under the `if err_handler is not None`?
In which scenario is it beneficial to reset the body if we haven't found any error handler?

CaselIT replied:
Yes, there should be no difference, since if we do not have an err_handler falcon does not manage the response and instead makes the wsgi serve handle it.
That should only happen on exception that do not inherit from `Exception` now, like `KeyboardInterrupt` or `SystemExit`, since we add an handler for `Exception`. This case should be very rare, so it's probably the same to have it inside or outside the if.
